Feature/tpa 0013 admin unapproved accounts

DeveloperX104 opened the pull request:
Un approved users are being rendered in the `admin/view/users/unapprove` endpoint.

DeveloperX104 committed:
unapproved users are being rendered

DeveloperX104 committed:
unapproved accounts are rendered in the admin side

DeveloperX104 committed:
removed selected unapproved accounts table

2kama commented on src/pages/AdminPages/UnapprovedAccounts.js:
```diff
@@ -0,0 +1,53 @@
+    )
+}
+
+export default UnapprovedAccounts
```
I feel we should follow a pattern where every page is made a folder with an `index.jsx` file in it. Then subsequent component that will only be utilized by that page will be added to the same folder.

Make this `UnapprovedAccounts/index.jsx`

DeveloperX104 replied:
Alright, thanks!

2kama commented on src/store/sagas/requests/user.js:
```diff
@@ -42,4 +42,10 @@ export const requestGetUserPrivateData = () => {
+}
+
+export const getUnapprovedUsers = () => {
+    return db.collection('newUser').get().then((users) => {
+        return users.docs.map((user) => user.data())
+    })
```
Remember I said `newUser` is not a reliable collection. Instead search for users in the `users` collection whole have their `isApproved` field set to false in the `private/info` subcollection. Then use their `email` to search for their document in the `newUser` collection

DeveloperX104 replied:
ooh okay.

DeveloperX104 replied:
So after looking at it again, I feel the we shouldn't create anything in the `users` collection until the account is approved, it makes it easier and better to work with. So when the user signs up, they have their data in the `newUser` collection then when the account is approved the data is transferred to the `users` collection

2kama replied:
Again, that can't work, remember `newUser` gets populated even if the register fails.... that means there could be data in newUser that doesn't belong to any registered user. It's an unreliable collection.

2kama commented on src/pages/AdminPages/UnapprovedAccounts.js:
```diff
@@ -0,0 +1,53 @@
+                            headers={['firstName', 'lastName', 'email', 'phone', 'actions']}
+                            data={unapprovedUsers.map(user => (
+                                {
+                                    ...user, 
```
We might have to rework on this. The admin has to be able to able to set some properties for the user before approving. The client doesn't want to just approving all users at once. he wants to be able to read through each registration and decide to approve or deny.

DeveloperX104 replied:
Yh, so here the user can click on the approve button which opens a modal, the admin can now set properties inside that modal. I'd need you to give me the data that the admin will be setting here also.

2kama replied:
`role` and `ROI` and `assignedTrader` ...that's all i can think of now.

`assignedTrader` is supposed to be a select field that shows the list of all the traders registered on the app. Then the admin can pick one of the Traders to assign to the user.

NOTE: if the account is made an admin or a trader; `ROI` and `assignedTrader` will not be needed....these fields will be hidden bcos an admin and Trader cannot have an investment portfolio, so ROI is useless and since they don't have an investment portfolio, they can't be assigned to a Trader

Then once, set `isApproved` to `true`, fix the data from `newUser` collection into their respective field in the `user` collection and delete that document from the `newUser` collection.

forget about emailVerification, once the user is approved, the hash page changes to the user clicking a button to send an email verification link (this is already done)

DeveloperX104 replied:
Alright thanks.

2kama commented on src/pages/AdminPages/UnapprovedAccounts.js:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    useEffect(() => {
+        dispatch(getUnapprovedUsers())
```
This dispatch call is running multiple times... This should only query the database once on page load. To update the page, the page has to be reloaded

2kama replied:
![Screen Shot 2021-07-22 at 11 17 11 PM](https://user-images.githubusercontent.com/6509866/126733875-5ce5f3bc-3988-494c-b995-1855e5181d66.png)

2kama replied:
And this is a very bad technique for firebase... Remember you get billed on document reads?
![Screen Shot 2021-07-22 at 11 21 24 PM](https://user-images.githubusercontent.com/6509866/126734065-2f194eec-1c51-4775-9714-d87fab0e5a5a.png)

2kama replied:
The spike there, is when you were running this code... 

If you were trying to give a realtime view of the collection, a `snapshot` is a better way to do it. That way we only get a document read when a document is added or removed from the collection.

DeveloperX104 replied:
Thanks, will fix this.

2kama commented on src/store/sagas/requests/user.js:
```diff
@@ -42,4 +42,10 @@ export const requestGetUserPrivateData = () => {
     return db.doc(`users/${auth.currentUser.uid}`).get()
+}
+
+export const getUnapprovedUsers = () => {
```
Admin getting unapproved users is not a user specific operation... a new file should be created for these. Maybe call it `adminQueryUsers.js`

DeveloperX104 replied:
ok

2kama commented on src/store/sagas/handlers/user.js:
```diff
@@ -76,4 +77,18 @@ export function* handleVerifyUser(action) {
+}
+
+
+export function* handleGetUnapprovedUsers(action) {
```
Admin getting unapproved users is not a user specific operation... a new file should be created for these. Maybe call it `adminQueryUsers.js` [Part II]

DeveloperX104 replied:
Alright, thanks.

2kama commented on src/store/sagas/handlers/user.js:
```diff
@@ -76,4 +77,18 @@ export function* handleVerifyUser(action) {
+        ))
+        
+    } catch (err) {
+        console.log(err)
```
Set this to push an alert if there's an error.

DeveloperX104 replied:
Ok

DeveloperX104 committed:
clicking on the approve button opens a modal to set the role of the user and approve the user.

2kama commented on src/pages/AdminPages/UnapprovedAccounts.js:
```diff
@@ -0,0 +1,115 @@
+            isAffiliate : false,
+            isTrader : false
+        }
+        if (e.target.value === 'admin') {
+            data.isAdmin = true;
+        } 
+        if (e.target.value === 'superadmin') {
+            data.isSuperAdmin = true;
+        }
+        if (e.target.value === 'user') {
+            data.isUser = true
+        }
+        if (e.target.value === 'affiliate') {
+            data.isAffiliate = true
+        }
+        if (e.target.value === 'trader') {
+            data.isTrader = true
+        }
```
1. You are presuming that the admin will not first click / accidentally choose something else before choosing the right role...You need to set the other roles to false when he picks  any role.

2. superadmin is also an admin so both will be true when superadmin is chosen. an affiliate is also a user so both will be true when affiliate is chosen

DeveloperX104 replied:
Alright, thanks for this.

DeveloperX104 committed:
update `setRole`

DeveloperX104 committed:
error notification

DeveloperX104 committed:
fixed minro changes

DeveloperX104 committed:
unapproved data is nor successfully rendered.

2kama commented on src/pages/AdminPages/UnapprovedAccounts/index.js:
```diff
@@ -0,0 +1,136 @@
+                                        <p>Email: {selectedUser.email}</p>
+                                        <p>Phone Number: {selectedUser.phone}</p>
+                                        role: <select onChange={setRole}>
+                                            <option defaultChecked value="select">select</option>
```
The default value will be `user`

DeveloperX104 committed:
updated modal

DeveloperX104 committed:
updated modal

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -0,0 +1,16 @@
+const db = firebase.firestore()
+
+export const getUnapprovedUsers = async () => {
+    let data = await db.collection('users').get()
+    let unapprovedUsers = []
+    for (let user of data.docs) {
+        let privateInfo = await db.doc(`users/${user.id}/private/info`).get()
+        if (!privateInfo.data().isApproved) {
+            let userData = await db.doc(`newUser/${privateInfo.data().email}`).get()
+             unapprovedUsers.push(userData.data())
```
You have to get `users` that have `isApproved` set to `false`, not all the `users`. There is a `where` method in firebase which you can use to filter document gets. Imagine having only 1 unapproved user and you'd have to get over 500 documents just for that 1 user. Pay Per document read

DeveloperX104 replied:
The `user` collection itself doesn't have a `isApproved` field.

2kama replied:
You wanna check that again? check the `cloud-functions` folder in the `functions` folder. Or check the firestore database

2kama replied:
???

DeveloperX104 replied:
Ooh thanks, just understamding what you meant here

DeveloperX104 replied:
I'm having issue trying to chain the query

2kama replied:
What's the issue?

2kama replied:
https://firebase.google.com/docs/firestore/query-data/queries#collection-group-query

DeveloperX104 replied:
Don't you think the `isApproved`  field should be inside the `users` collection directly and not in a subcollection.

2kama replied:
all the data that is inside the `private/info` document are data that the user cannot edit. the data that is directly in the `users` document is the data the user can edit.

We wouldn't want the user to be able to edit his `wallet`, `approval` status, `uid`, `role` etc...

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -0,0 +1,16 @@
+             unapprovedUsers.push(userData.data())
+        }
+    }
+    return unapprovedUsers.reverse()
```
This is not a blocker, but why do you need to reverse?

DeveloperX104 replied:
I feel having the recent data at the top is better.

2kama replied:
Well, there really is no need for it. The data you're getting is arranged in alphabetical order of the `uid` ... and u know `uid` is a mayhem of alpha-numeric irrespective of who came first.

If you were to sort the data somehow (maybe by `createdAt` field), the oldest entry should be first...not the latest. But even this will need you to start creating composite indexes on firebase...so...on the long run, not needed.

but again, this is no blocker at all.

DeveloperX104 replied:
Thanks for the heads up!

DeveloperX104 committed:
updated modal

DeveloperX104 committed:
added label

2kama commented on src/pages/AdminPages/UnapprovedAccounts/index.js:
```diff
@@ -0,0 +1,143 @@
+                                    ...user,
+                                    'approve':  <>
+                                        <button onClick={() => {setShowModal(true); setSelectedUser(user)}}>Approve</button>
+                                        <BaseModal 
+                                            title={`Approve ${selectedUser.email}'s account`}
+                                            show={showModal}
+                                            close={() => {setShowModal(false)}}
+                                            doneText="Approve" 
+                                            closeText="Cancel"
+                                            onDone={() => approveAccount(user)}
+                                        >
+                                        <p>Are you sure you want to approve {user.email}'s account?</p>
+                                        <p>First Name: {selectedUser.firstName}</p>
+                                        <p>Last Name: {selectedUser.lastName}</p>
+                                        <p>Email: {selectedUser.email}</p>
```
Let's use `Formik` like I did in Login and Register to process all our forms... Thanks

DeveloperX104 replied:
Alright

DeveloperX104 committed:
fixed minor bug

DeveloperX104 committed:
fetching data of traders in the modal

DeveloperX104 committed:
.

DeveloperX104 committed:
fixed modal header bug

DeveloperX104 committed:
modal is now complete and working fine!

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -0,0 +1,28 @@
+}
+
+export const getAllTraders = async () => {
+    let data = await db.collection('users').get()
+    let traders = []
+    for (let user of data.docs) {
+        let privateInfo = await db.doc(`users/${user.id}/private/info`).get()
+        if (privateInfo.data().role.isTrader) {
+             traders.push({...user.data(), ...privateInfo.data()})
+        }
+    }
+    return traders
```
You need to fix this... use a `where` method in firebase so you're only reading documents that you need and not all the documents

DeveloperX104 committed:
better query

DeveloperX104 committed:
better query

2kama commented on src/pages/AdminPages/UnapprovedAccounts/index.js:
```diff
@@ -0,0 +1,163 @@
+import React, {useEffect, useState} from 'react'
```
This component is doing a lot. Let's trim this down.

Make each `unapproved user` row a component...that way the logic won't be so complex. And try utilizing the `Form Components` built. I've used it a couple of times so you can get an idea of how it works. It'll make Form functions neater and more readable

DeveloperX104 replied:
Thanks, I'd do this now

DeveloperX104 committed:
break down into smaller components

DeveloperX104 committed:
`FormSelect` component updated and used.

2kama commented on src/components/Modal/BaseModal.js:
```diff
@@ -0,0 +1,26 @@
+        <Modal
+            show={show}
+            onHide={close}
+            size="lg"
```
let this be a prop.... we might need a `md` or `sm` modal

2kama replied:
add a `size` props

2kama commented on src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:
```diff
@@ -0,0 +1,145 @@
+    )
+}
+
+export default SelectedUnapprovedUserModal
```
I still have issues with the complexity of this file

DeveloperX104 replied:
Most of what is in this file are functions, I've tried making the file smaller

2kama commented on src/components/Modal/BaseModal.js:
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import { Modal, Button } from 'react-bootstrap';
+
+const BaseModal = ({show, close, title, children, closeText}) => {
```
What was your reason for updating the props of your modal?

2kama replied:
Not that I don't like this...it looks cleaner

DeveloperX104 replied:
I removed the `doneText` and the `onDone` props as I'm using form here so I do not need those props

DeveloperX104 committed:
account approval now works like charm!

DeveloperX104 committed:
removed a userDelete cloud function

DeveloperX104 committed:
.

DeveloperX104 committed:
merged

DeveloperX104 committed:
.

DeveloperX104 committed:
.

DeveloperX104 committed:
deleting unapproved users

2kama committed:
Optimized and Cleaned Up

2kama committed:
Just updating the FormSelect component

DeveloperX104 committed:
delete account works in the table component

DeveloperX104 committed:
done?

DeveloperX104 committed:
Merge branch 'master' of https://github.com/2kama/tpa into feature/TPA-0013-admin-unapproved-accounts

2kama commented on src/pages/AdminPages/UnapprovedAccounts/UnapprovedUsers.js:
```diff
@@ -3,7 +3,8 @@ import { useSelector, useDispatch, shallowEqual } from 'react-redux'
+import { disableButton } from '../../../store/reducers/buttonState';
 
 const UnapprovedUsers = () => {
 
```
Why are we back to using `UnapprovedUsers` ?

DeveloperX104 committed:
.

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -26,20 +26,21 @@ export const getAllTraders = async () => {
 
-export const approveUser = async (data) => {
-    await db.doc(`users/${data.id}/private/info`).update({
+export const alterUser = async (data) => {
```
This function shouldn't be an async function. We really don't need to worry if the `firstName`, `lastName` and `phone` is updated bcos the user can still do this himself if something wrong happens on the admin end.

But we are particular about the `private` collection being updated bcos that is the whole purpose of the update user process.... so we need the promise returned from updating the private collection --- we need to know if it failed or pass.

with this technique, we'll never know bcos you're not even returning the promise to the `handler`.

2kama replied:
The implementation from my last commit on this function is better aligned to this.

2kama replied:
```
export const alterUser = (data) => {

    db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })
    
    return db.doc(`users/${data.uid}/private/info`).update({
        role: data.role,
        isApproved: true,
        ROI: data.roi < 0 ? 0 : data.roi,
        assignedTrader: data.role.isUser ? data.assignedTrader : "",
        affiliateCode: data.role.isAffliate ? data.affiliateCode : "",
        affiliate: data.affiliate
    })
}
```

Change the whole func to this...

2kama commented on functions/cloud_functions/user.js:
```diff
@@ -56,4 +56,24 @@ exports.userCreated = functions.auth.user().onCreate(async user => {
+
+exports.deleteUnapprovedUser = functions.firestore.document("/users/{userId}/private/info").onDelete(async (data, context) => {
+    try {
+        await admin.firestore().doc(`newUser/${data.data().email}`).delete();
```
Delete this line... we are not doing anything to `newUser` but populating and reading it for approvals

2kama commented on functions/index.js:
```diff
@@ -13,6 +13,6 @@ const userCreated = require('./cloud_functions/userCreated')
-exports.userCreated = userCreated.userCreated
\ No newline at end of file
+exports.userCreated = user.userCreated;
+exports.userApproved = user.userApproved;
```
Not a needed cloud function.... `userApproved`

2kama commented on src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:
```diff
@@ -0,0 +1,131 @@
+        firstName: Yup.string().required().min(3).label("First Name"),
+        lastName: Yup.string().required().min(3).label("Last Name"),
+        phone: Yup.string().required().label("Phone Number"),
+        roi: selectedAlterUser.role && selectedAlterUser.role.isUser && Yup.number().required().min(0).max(100).label("ROI"),
+        affiliateCode: selectedAlterUser.role && selectedAlterUser.role.isAffiliate && Yup.string().required().length(6).label("Affiliate Code"),
+        assignedTrader: selectedAlterUser.role && selectedAlterUser.role.isUser && Yup.string().required().label("Assigned Trader"),
```
Take these off... Not necessary.... They aren't a `mustFill` field.

2kama commented on functions/cloud_functions/user.js:
```diff
@@ -56,4 +56,24 @@ exports.userCreated = functions.auth.user().onCreate(async user => {
\ No newline at end of file
+})
+
+exports.userApproved = functions.firestore.document("/users/{userId}/private/info").onUpdate(async (change, context) => {
+    try {
+        await admin.firestore().doc(`newUser/${change.before.data().email}`).delete();
+    } catch (error) {
+        console.log(error.message)
+    }
+});
```
we don't need this cloud function...pay no attention to `newUser` collection, it's just a dumping group.

Besides, anytime an admin updates a user, even though he wasn't approving a user, this function will run...so it's a waste of read and write activities.

2kama commented on src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:
```diff
@@ -0,0 +1,131 @@
+            dispatch(setAlterUsers(
+                alterUsers.filter(user => user.uid !== selectedAlterUser.uid)
+            ))
+            dispatch(alterUserDispatch(selectedAlterUser))
```
What is this for?

DeveloperX104 replied:
This is the dispatch function that actually calls the generator that approves the user account.

2kama commented on src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:
```diff
@@ -0,0 +1,131 @@
+import { superAdminSettableRoles, adminSettableRoles } from '../../../utils/userRoles'
+import { alterUser as alterUserReducer, setAlterUsers, setSelectedAlterUser } from '../../../store/reducers/adminQuery';
+import { getIndexOfK } from '../../../utils/helperFunctions'
+import { alterUser as alterUserDispatch } from '../../../store/reducers/adminQuery';
```
Why are we importing this twice?

DeveloperX104 committed:
...

DeveloperX104 committed:
...

DeveloperX104 committed:
ok

DeveloperX104 committed:
ok

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -0,0 +1,56 @@
+    data.affiliateCode && (updatedData.affiliateCode = data.affiliateCode)
+    data.referralCode && (updatedData.referralCode = data.referralCode)
+
+    db.doc(`users/${data.uid}/private/info`).update(updatedData)
+    
+    return db.doc(`users/${data.uid}`).update({
+        firstName: data.firstName,
+        lastName: data.lastName,
+        phone: data.phone
+    })
```
Change to this...

```
db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })

return db.doc(`users/${data.uid}/private/info`).update(updatedData)
```

2kama replied:
We want the promise from updating the `private/info` not the `users` document

DeveloperX104 committed:
referral code field

DeveloperX104 committed:
.

2kama commented on src/pages/AdminPages/UnapprovedAccounts/SelectedUserModal.js:
```diff
@@ -0,0 +1,138 @@
+            }))
+        }
+        closeModal()
+    }
```
Return this function back to how it was previously

DeveloperX104 committed:
.

2kama commented on src/store/sagas/requests/adminQueryUsers.js:
```diff
@@ -0,0 +1,56 @@
+    return traders
+}
+
+export const alterUser = (data) => {
+    const updatedData = {
+        isApproved: true
+    }
+    data.roi && (updatedData.roi = data.roi)
+    data.assignedTrader && (updatedData.assignedTrader = data.assignedTrader)
+    data.role && (updatedData.role = data.role)
+    data.affiliateCode && (updatedData.affiliateCode = data.affiliateCode)
+    data.referralCode && (updatedData.referralCode = data.referralCode)
+
+    
+    db.doc(`users/${data.uid}`).update({
+        firstName: data.firstName,
+        lastName: data.lastName,
+        phone: data.phone
+    })
+    return db.doc(`users/${data.uid}/private/info`).update(updatedData)
+}
```
Change all these to : 


```
export const alterUser = (data) => {

    db.doc(`users/${data.uid}`).update({
        firstName: data.firstName,
        lastName: data.lastName,
        phone: data.phone
    })
    
    return db.doc(`users/${data.uid}/private/info`).update({
        role: data.role,
        isApproved: true,
        ROI: data.roi < 0 ? 0 : data.roi,
        assignedTrader: data.role.isUser ? data.assignedTrader : "",
        affiliateCode: data.role.isAffliate ? data.affiliateCode : "",
        referralCode: data.referralCode
    })
}
```

2kama replied:
Over here

2kama commented on functions/cloud_functions/user.js:
```diff
@@ -39,7 +38,8 @@ exports.userCreated = functions.auth.user().onCreate(async user => {
         assignedTrader : "",
         isApproved : false,
-        affiliate : ""
+        affiliate : "",
```
change to `referralCode`

DeveloperX104 committed:
.

DeveloperX104 committed:
.

2kama committed:
Return Promise

Try not to be eager to use async functions... you needed to only return the promise so that your status check in the handler will be triggered.
